### ISSUE TITLE
Add distributed storage layer for containers

### ISSUE THREAD

**Issue by BrianHicks** [CiscoCloud/microservices-infrastructure#567]
Some ideas:

 - Flocker (but it only supports certain clouds at the moment, and *experimental* ZFS clustering)
 - NFS
 - GlusterFS

In any case, it would be helpful for a number of tasks to be able to have persistent storage that *moved around the cluster* when needed. Flocker in particular is well-suited for that task.

**Comment by tgermain** [comment]
+1.

 I've already have a GlusterFS setup with ansible but it's very dependent of your partition scheme (plus I don't work with cloud provider but directly on VMs).

I think we need to discuss if you want to include the whole glusterFS setup (summarize from [glusterFS quickstart](http://gluster.readthedocs.org/en/latest/Quick-Start-Guide/Quickstart/) ):

1.  Create the back-end for glusterFS : the partitions used for the bricks of glusterfs ( how many bricks ? which size ? ... ).
2.  Create the glusterFS volume (how many replicas ? ...)
3.  Mount the glusterFS volume (where ? ...).

I also think that you want this glusterFS volume to be mounted on every resource node but do you also want them(the resource nodes) to be part of the trusted pool of glusterFS ?

Basically, step 1 can be very tricky to generalize as you deal with partitions.

**Comment by kenjones-cisco** [comment]
I took a very similar approach with my instance where I created the cinder volumes using terraform to attach to my VMs. And then using ansible created the logical volume group and logical volumes, mounted to the block device, and then used GlusterFS to created replicated volumes across each worker node in the cluster, and mounted those as directories.

Then when my database starts up on any node in the cluster, the storage is there and in sync.

I agree, doing that generically will be tricky for sure. Flocker could play a role as the plugin approach gets less experimental. Since I was going production, I was looking for something a bit less bleeding edge.

If it would be helpful, I can share the ansible and terraform components.

**Comment by abn** [comment]
@kenjones-cisco +1; I reckon your approach makes a lot of sense and especially considering production use cases. It'd be quite helpful if you could share playbooks and terraform specifications.

**Comment by BrianHicks** [comment]
@kenjones-cisco @abn I've got a very similar approach going in feature/glusterfs. Right now the challenge is getting it to work on all the clouds we support. If y'all want to check out the current configuration there, I'd appreciate any feedback you have.
